cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationNotificationSubscriberService.java

```diff
@@ -0,0 +1,353 @@
+      ImmutablePair<String, Notification> messagePair = messages.next();
+      List<Runnable> runnables =
+          processNotification(
+              messagePair.getFirst().getBytes(StandardCharsets.UTF_8),
+              messagePair.getSecond(),
+              structuredTableContext);
+      tasks.addAll(runnables);
+    }
+
+    // Only add post-processing tasks if all messages are processed. If there is exception in the
+    // processNotification, messages will be replayed.
+    this.tasks.addAll(tasks);
+  }
+
+  @Override
+  protected void postProcess() {
+    Runnable task = tasks.poll();
+    while (task != null) {
+      task.run();
+      task = tasks.poll();
+    }
+  }
+
+  /**
+   * Process a {@link Notification} received from TMS.
+   *
+   * @param messageIdBytes the raw message id in the TMS for the notification
+   * @param notification the {@link Notification} to process
+   * @param context context to get the table for operations
+   * @return a {@link List} of {@link Runnable} tasks to run after the transactional processing of
+   *     the whole messages batch is completed
+   * @throws Exception if failed to process the given notification
+   */
+  @VisibleForTesting
+  List<Runnable> processNotification(
+      byte[] messageIdBytes,
+      Notification notification,
+      StructuredTableContext context)
+      throws Exception {
+    OperationRunStore runStore = new OperationRunStore(context);
+    List<Runnable> result = new ArrayList<>();
+    OperationNotification operationNotification;
+
+    try {
+      operationNotification = OperationNotification.fromNotification(
+          notification);
+    } catch (IllegalArgumentException | JsonSyntaxException e) {
+      LOG.warn("Got invalid operation notification", e);
+      return result;
+    }
+
+    handleOperationEvent(
+        operationNotification,
+        messageIdBytes,
+        runStore,
+        result
+    );
+
+    return result;
+  }
+
+  private void handleOperationEvent(
+      OperationNotification notification,
+      byte[] messageIdBytes,
+      OperationRunStore runStore,
+      List<Runnable> runnables
+  ) throws Exception {
+    LOG.trace("Processing operation status notification: {}", notification);
+    OperationRunId runId = notification.getRunId();
+    OperationRunDetail runDetail;
+    try {
+      runDetail = runStore.getOperation(runId);
+      if (!runDetail.getRun().getStatus().canTransitionTo(notification.getStatus())) {
+        LOG.debug(
+            "Ignoring unexpected request to transition operation {} from {} state to "
+                + "{} state.",
+            runId, runDetail.getRun().getStatus(), notification.getStatus());
+        return;
+      }
+    } catch (OperationRunNotFoundException e) {
+      LOG.debug(String.format("Ignoring message for non existent operation %s", runId));
+      return;
+    }
+    switch (notification.getStatus()) {
+      case PENDING:
+        statePublisher.publishStarting(runId);
```

Comment:
The PENDING and STARTING state was there for two reasons
- Make it similar to program run states so that it can be convereged later
- We can introduce a cluster provisioning like layer later similar to program to have the capability of creating a new worker in case all existing workers are busy.

I will remove the STARTING status as it does not have any usage right now